README.md
stamped_velocity_smoother
=========================

This package handles provides a simple velocity smoother using a TwistStamped message.

The code is based on the yocs_velocity_smoother ( 0.8.1 version 2016-03-16) from https://github.com/yujinrobot/yujin_ocs.git.

This version uses boost threads to reduce dependencies on external libraries.






CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package stamped_velocity_smoother
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.0.1 version based on 0.8.1 version (2016-03-16) from https://github.com/yujinrobot/yujin_ocs.git
Converted to use boost thread to reduce dependency
Use TwistStamped for output, and allow different input types
------------------

* Initial version.


test/test_translational_input.py
#!/usr/bin/env python
import roslib
roslib.load_manifest('stamped_velocity_smoother')
import rospy

import os
import sys
import time
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
'''
  Varied translational input for a velocity smoother test.
'''

STATE_RAMP_UP = 0
STATE_RAMP_LEVEL = 1
STATE_RAMP_DOWN = 2
STATE_ZERO = 3
STATE_UP = 4
STATE_DOWN = 5
STATE_UP_AGAIN = 6
STATE_NOTHING = 7

def main():
    rospy.init_node("test_velocity_smoother_input")
    cmd_vel_publisher = rospy.Publisher("~cmd_vel", Twist)
    odom_publisher = rospy.Publisher("~odom", Odometry)
    param = {}
    param['velocity_maximum'] = rospy.get_param("~velocity_maximum", 0.50)
    param['ramp_increment'] = rospy.get_param("~ramp_increment", 0.02)
    rospy.loginfo("Test Input : ramp increment [%f]",param['ramp_increment'])
    param['ramp_decrement'] = rospy.get_param("~ramp_decrement", 0.02)
    rospy.loginfo("Test Input : ramp decrement [%f]",param['ramp_decrement'])
    cmd_vel = Twist()
    cmd_vel.linear.x = 0
    cmd_vel.linear.y = 0
    cmd_vel.linear.z = 0
    cmd_vel.angular.x = 0
    cmd_vel.angular.y = 0
    cmd_vel.angular.z = 0
    odom = Odometry()
    odom.header.frame_id = "base_link"
    odom.pose.pose.position.x = 0.0
    odom.pose.pose.position.y = 0.0
    odom.pose.pose.position.z = 0.0
    odom.pose.pose.orientation.x = 0.0
    odom.pose.pose.orientation.y = 0.0
    odom.pose.pose.orientation.z = 0.0
    odom.pose.pose.orientation.w = 1.0

    odom.pose.covariance[0]  = 0.1
    odom.pose.covariance[7]  = 0.1
    odom.pose.covariance[35] = 0.2
    odom.pose.covariance[14] = 10.0
    odom.pose.covariance[21] = 10.0
    odom.pose.covariance[28] = 10.0

    odom.twist.twist.linear.x = 0.0
    odom.twist.twist.linear.y = 0.0
    odom.twist.twist.linear.z = 0.0
    odom.twist.twist.angular.x = 0.0
    odom.twist.twist.angular.y = 0.0
    odom.twist.twist.angular.z = 0.0
    state = STATE_RAMP_UP
    count = 0
    count_max = 100
    publish = True
    #period = 0.01
    timer = rospy.Rate(100) # 10hz
    rospy.loginfo("Test Input : STATE_RAMP_UP")
    while not rospy.is_shutdown():
        if state == STATE_RAMP_UP:
            cmd_vel.linear.x = cmd_vel.linear.x + param['ramp_increment']
            if cmd_vel.linear.x >= param['velocity_maximum']:
                state = STATE_RAMP_LEVEL
                count = 0
                rospy.loginfo("Test Input : STATE_RAMP_UP -> STATE_RAMP_LEVEL")
        elif state == STATE_RAMP_LEVEL:
            if count > count_max: # 0.5s
                state = STATE_RAMP_DOWN
                count = 0
                rospy.loginfo("Test Input : STATE_RAMP_LEVEL -> STATE_RAMP_DOWN")
            else:
                count = count + 1
        elif state == STATE_RAMP_DOWN:
            cmd_vel.linear.x = cmd_vel.linear.x - param['ramp_decrement']
            if cmd_vel.linear.x <= 0.0:
                cmd_vel.linear.x = 0.0
                state = STATE_ZERO
                count = 0
                rospy.loginfo("Test Input : STATE_RAMP_DOWN -> STATE_ZERO")
        elif state == STATE_ZERO:
            if count > count_max: # 0.5s
                state = STATE_UP
                cmd_vel.linear.x = param['velocity_maximum']
                count = 0
                rospy.loginfo("Test Input : STATE_ZERO -> STATE_UP")
            else:
                count = count + 1
        elif state == STATE_UP:
            if count > count_max: # 0.5s
                state = STATE_DOWN
                cmd_vel.linear.x = 0.0
                count = 0
                rospy.loginfo("Test Input : STATE_UP -> STATE_DOWN")
            else:
                count = count + 1
        elif state == STATE_DOWN:
            if count > count_max: # 0.5s
                #state = STATE_UP_AGAIN
                #cmd_vel.linear.x = param['velocity_maximum']
                #rospy.loginfo("Test Input : STATE_DOWN -> STATE_UP_AGAIN")
                state = STATE_RAMP_UP
                cmd_vel.linear.x = 0.0
                rospy.loginfo("Test Input : STATE_DOWN -> STATE_RAMP_UP")
                count = 0
            else:
                count = count + 1
        elif state == STATE_UP_AGAIN:
            if count > count_max: # 0.5s
                state = STATE_NOTHING
                count = 0
                publish = False
                rospy.loginfo("Test Input : STATE_UP_AGAIN -> STATE_NOTHING")
            else:
                count = count + 1
        elif state == STATE_NOTHING:
            if count > count_max: # 0.5s
                state = STATE_RAMP_UP
                cmd_vel.linear.x = 0.0
                count = 0
                publish = True
                rospy.loginfo("Test Input : STATE_NOTHING -> STATE_RAMP_UP")
            else:
                count = count + 1
        if publish:
            odom.twist.twist.linear.x = cmd_vel.linear.x
            cmd_vel_publisher.publish(cmd_vel)
        else:
            # How to fake it when it's not publishing a cmd velocity? Up to the velocity controller there
            odom.twist.twist.linear.x = cmd_vel.linear.x
        odom.header.stamp = rospy.Time().now()
        odom_publisher.publish(odom)
        timer.sleep()

if __name__ == "__main__":
  main()


include/stamped_velocity_smoother/velocity_smoother_nodelet.hpp
/**
 * @file /include/stamped_velocity_smoother/velocity_smoother_nodelet.hpp
 *
 * @brief Velocity smoother interface
 *
 * License: BSD
 *   https://raw.github.com/CNURobotics/stamped_velocity_smoother/LICENSE
 **/
/*****************************************************************************
 ** Ifdefs
 *****************************************************************************/

#ifndef STAMPED_VELOCITY_SMOOTHER_HPP_
#define STAMPED_VELOCITY_SMOOTHER_HPP_

/*****************************************************************************
 ** Includes
 *****************************************************************************/

#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/TwistStamped.h>
#include <boost/thread.hpp>
#include <boost/thread/mutex.hpp>

/*****************************************************************************
** Namespaces
*****************************************************************************/

namespace stamped_velocity_smoother {

/*****************************************************************************
** VelocitySmoother
*****************************************************************************/

class VelocitySmoother
{
public:
  VelocitySmoother(const std::string &name);

  ~VelocitySmoother()
  {
    if (dynamic_reconfigure_server != NULL)
      delete dynamic_reconfigure_server;
  }

  bool init(ros::NodeHandle& nh);
  void spin();
  void shutdown() { shutdown_req = true; };

private:
  enum RobotFeedbackType
  {
    NONE,
    ODOMETRY,
    COMMANDS
  } robot_feedback;  /**< What source to use as robot velocity feedback */

  std::string name;
  bool quiet;        /**< Quieten some warnings that are unavoidable because of velocity multiplexing. **/
  double speed_lim_v, accel_lim_v, decel_lim_v;
  double speed_lim_w, accel_lim_w, decel_lim_w;
  double decel_factor;

  double frequency;

  geometry_msgs::TwistStamped           last_cmd_vel;
  geometry_msgs::TwistStamped           current_vel;
  geometry_msgs::TwistStamped           target_vel;
  geometry_msgs::TwistStamped           last_target_vel_msg;

  boost::mutex                          mutex;

  bool                 shutdown_req; /**< Shutdown requested by nodelet; kill worker thread */
  bool                 input_active;
  double                cb_avg_time;
  ros::Time            last_cb_time;
  std::vector<double> period_record; /**< Historic of latest periods between velocity commands */
  unsigned int             pr_next; /**< Next position to fill in the periods record buffer */

  ros::Subscriber odometry_sub;    /**< Current velocity from odometry */
  ros::Subscriber current_vel_sub; /**< Current velocity from commands sent to the robot, not necessarily by this node */
  ros::Subscriber raw_in_vel_sub;  /**< Incoming raw velocity commands */
  ros::Publisher  smooth_vel_pub;  /**< Outgoing smoothed velocity commands */

  void twistCB(const geometry_msgs::Twist::ConstPtr& msg);
  void twistStampedCB(const geometry_msgs::TwistStamped::ConstPtr& msg);
  void velocityTarget(const geometry_msgs::Twist& msg);

  void robotVelCB(const geometry_msgs::TwistStamped::ConstPtr& msg);
  void odometryCB(const nav_msgs::Odometry::ConstPtr& msg);

  double sign(double x)  { return x < 0.0 ? -1.0 : +1.0; };

  double median(std::vector<double> values) {
    // Return the median element of an doubles vector
    nth_element(values.begin(), values.begin() + values.size()/2, values.end());
    return values[values.size()/2];
  };

  /*********************
  ** Dynamic Reconfigure
  **********************/
  dynamic_reconfigure::Server<stamped_velocity_smoother::paramsConfig> *             dynamic_reconfigure_server;
  dynamic_reconfigure::Server<stamped_velocity_smoother::paramsConfig>::CallbackType dynamic_reconfigure_callback;
  void reconfigCB(stamped_velocity_smoother::paramsConfig &config, uint32_t unused_level);
};

} // stamped_namespace velocity_smoother

#endif /* STAMPED_VELOCITY_SMOOTHER_HPP_ */


src/velocity_smoother_nodelet.cpp
/**
 * @file /src/velocity_smoother_nodelet.cpp
 *
 * @brief Velocity smoother implementation.
 *
 * License: BSD
 *   https://raw.github.com/CNURobotics/stamped_velocity_smoother/LICENSE
 **/
/*****************************************************************************
 ** Includes
 *****************************************************************************/

#include <ros/ros.h>
#include <nodelet/nodelet.h>
#include <pluginlib/class_list_macros.h>

#include <dynamic_reconfigure/server.h>
#include <stamped_velocity_smoother/paramsConfig.h>

#include <boost/thread.hpp>

#include "stamped_velocity_smoother/velocity_smoother_nodelet.hpp"

/*****************************************************************************
 ** Preprocessing
 *****************************************************************************/

#define PERIOD_RECORD_SIZE    5
#define ZERO_VEL_COMMAND      geometry_msgs::Twist();
#define IS_ZERO_VEOCITY(a)   ((a.twist.linear.x == 0.0) && (a.twist.angular.z == 0.0))

/*****************************************************************************
** Namespaces
*****************************************************************************/

namespace stamped_velocity_smoother {

/*********************
** Implementation
**********************/

VelocitySmoother::VelocitySmoother(const std::string &name)
: name(name)
, quiet(false)
, shutdown_req(false)
, input_active(false)
, pr_next(0)
, dynamic_reconfigure_server(NULL)
{
};

void VelocitySmoother::reconfigCB(stamped_velocity_smoother::paramsConfig &config, uint32_t level)
{
  ROS_INFO("Reconfigure request : %f %f %f %f %f",
           config.speed_lim_v, config.speed_lim_w, config.accel_lim_v, config.accel_lim_w, config.decel_factor);

  speed_lim_v  = config.speed_lim_v;
  speed_lim_w  = config.speed_lim_w;
  accel_lim_v  = config.accel_lim_v;
  accel_lim_w  = config.accel_lim_w;
  decel_factor = config.decel_factor;
  decel_lim_v  = decel_factor*accel_lim_v;
  decel_lim_w  = decel_factor*accel_lim_w;
}

void VelocitySmoother::twistCB(const geometry_msgs::Twist::ConstPtr& msg)
{
    boost::unique_lock<boost::mutex> scoped_lock(mutex);
    last_target_vel_msg.header.stamp = ros::Time::now();
    velocityTarget(*msg);
}
void VelocitySmoother::twistStampedCB(const geometry_msgs::TwistStamped::ConstPtr& msg)
{
    boost::unique_lock<boost::mutex> scoped_lock(mutex);
    last_target_vel_msg.header.stamp = msg->header.stamp;
    velocityTarget(msg->twist);
}

void VelocitySmoother::velocityTarget(const geometry_msgs::Twist& msg)
{
  // Estimate commands frequency; we do continuously as it can be very different depending on the
  // publisher type, and we don't want to impose extra constraints to keep this package flexible
  if (period_record.size() < PERIOD_RECORD_SIZE)
  {
    period_record.push_back((ros::Time::now() - last_cb_time).toSec());
  }
  else
  {
    period_record[pr_next] = (ros::Time::now() - last_cb_time).toSec();
  }

  pr_next++;
  pr_next %= period_record.size();
  last_cb_time = ros::Time::now();

  if (period_record.size() <= PERIOD_RECORD_SIZE/2)
  {
    // wait until we have some values; make a reasonable assumption (10 Hz) meanwhile
    cb_avg_time = 0.1;
  }
  else
  {
    // enough; recalculate with the latest input
    cb_avg_time = median(period_record);
  }

  input_active = true;

  // Bound speed with the maximum values
  last_target_vel_msg.twist.linear.x  =
      msg.linear.x  > 0.0 ? std::min(msg.linear.x,  speed_lim_v) : std::max(msg.linear.x,  -speed_lim_v);
  last_target_vel_msg.twist.angular.z =
      msg.angular.z > 0.0 ? std::min(msg.angular.z, speed_lim_w) : std::max(msg.angular.z, -speed_lim_w);
}

void VelocitySmoother::odometryCB(const nav_msgs::Odometry::ConstPtr& msg)
{
  if (robot_feedback == ODOMETRY)
    current_vel.twist = msg->twist.twist;

  // ignore otherwise
}

void VelocitySmoother::robotVelCB(const geometry_msgs::TwistStamped::ConstPtr& msg)
{
  if (robot_feedback == COMMANDS)
    current_vel = *msg;

  // ignore otherwise
}

void VelocitySmoother::spin()
{
  double period = 1.0/frequency;
  ros::Rate spin_rate(frequency);

  ROS_WARN("Starting spin loop for velocity smoother ...");
  while (! shutdown_req && ros::ok())
  {
    {   // update the data from ROS thread
        boost::unique_lock<boost::mutex> scoped_lock(mutex);
        target_vel = last_target_vel_msg;
    }
    if ((input_active == true) && (cb_avg_time > 0.0) &&
        ((ros::Time::now() - last_cb_time).toSec() > std::min(3.0*cb_avg_time, 0.5)))
    {
      // Velocity input no active anymore; normally last command is a zero-velocity one, but reassure
      // this, just in case something went wrong with our input, or he just forgot good manners...
      // Issue #2, extra check in case cb_avg_time is very big, for example with several atomic commands
      // The cb_avg_time > 0 check is required to deal with low-rate simulated time, that can make that
      // several messages arrive with the same time and so lead to a zero median
      input_active = false;
      if (IS_ZERO_VEOCITY(target_vel) == false)
      {
        ROS_WARN_STREAM("Velocity Smoother : input got inactive leaving us a non-zero target velocity ("
              << target_vel.twist.linear.x << ", " << target_vel.twist.angular.z << "), zeroing...[" << name << "]");
        target_vel.header.stamp = ros::Time::now();
        target_vel.twist = ZERO_VEL_COMMAND;
      }
    }

    if ((robot_feedback != NONE) && (input_active == true) && (cb_avg_time > 0.0) &&
        (((ros::Time::now() - last_cb_time).toSec() > 5.0*cb_avg_time)     || // 5 missing msgs
          (std::abs(current_vel.twist.linear.x  - last_cmd_vel.twist.linear.x)  > 0.2) ||
          (std::abs(current_vel.twist.angular.z - last_cmd_vel.twist.angular.z) > 2.0)))
    {
      // If the publisher has been inactive for a while, or if our current commanding differs a lot
      // from robot velocity feedback, we cannot trust the former; rely on robot's feedback instead
      // TODO: current command/feedback difference thresholds are arbitrary; they should somehow
      // be proportional to max v and w...
      // The one for angular velocity is very big because is it's less necessary (for example the
      // reactive controller will never make the robot spin) and because the gyro has a 15 ms delay
      if ( !quiet ) {
        // this condition can be unavoidable due to preemption of current velocity control on
        // velocity multiplexer so be quiet if we're instructed to do so
        ROS_WARN_STREAM("Velocity Smoother : using robot velocity feedback " <<
                        std::string(robot_feedback == ODOMETRY ? "odometry" : "end commands") <<
                        " instead of last command: " <<
                        (ros::Time::now() - last_cb_time).toSec() << ", " <<
                        current_vel.twist.linear.x  - last_cmd_vel.twist.linear.x << ", " <<
                        current_vel.twist.angular.z - last_cmd_vel.twist.angular.z << ", [" << name << "]"
                        );
      }
      last_cmd_vel.twist = current_vel.twist;
      last_cmd_vel.header= current_vel.header;
    }

    geometry_msgs::TwistStampedPtr cmd_vel;

    if ((target_vel.twist.linear.x  != last_cmd_vel.twist.linear.x) ||
        (target_vel.twist.angular.z != last_cmd_vel.twist.angular.z))
    {
      // Try to reach target velocity ensuring that we don't exceed the acceleration limits
      cmd_vel.reset(new geometry_msgs::TwistStamped());
      cmd_vel->header = target_vel.header;
      cmd_vel->twist  = target_vel.twist;

      double v_inc, w_inc, max_v_inc, max_w_inc;

      v_inc = target_vel.twist.linear.x - last_cmd_vel.twist.linear.x;
      if ((robot_feedback == ODOMETRY) && (current_vel.twist.linear.x*target_vel.twist.linear.x < 0.0))
      {
        // countermarch (on robots with significant inertia; requires odometry feedback to be detected)
        max_v_inc = decel_lim_v*period;
      }
      else
      {
        max_v_inc = ((v_inc*target_vel.twist.linear.x > 0.0)?accel_lim_v:decel_lim_v)*period;
      }

      w_inc = target_vel.twist.angular.z - last_cmd_vel.twist.angular.z;
      if ((robot_feedback == ODOMETRY) && (current_vel.twist.angular.z*target_vel.twist.angular.z < 0.0))
      {
        // countermarch (on robots with significant inertia; requires odometry feedback to be detected)
        max_w_inc = decel_lim_w*period;
      }
      else
      {
        max_w_inc = ((w_inc*target_vel.twist.angular.z > 0.0)?accel_lim_w:decel_lim_w)*period;
      }

      // Calculate and normalise vectors A (desired velocity increment) and B (maximum velocity increment),
      // where v acts as coordinate x and w as coordinate y; the sign of the angle from A to B determines
      // which velocity (v or w) must be overconstrained to keep the direction provided as command
      double MA = sqrt(    v_inc *     v_inc +     w_inc *     w_inc);
      double MB = sqrt(max_v_inc * max_v_inc + max_w_inc * max_w_inc);

      double Av = std::abs(v_inc) / MA;
      double Aw = std::abs(w_inc) / MA;
      double Bv = max_v_inc / MB;
      double Bw = max_w_inc / MB;
      double theta = atan2(Bw, Bv) - atan2(Aw, Av);

      if (theta < 0)
      {
        // overconstrain linear velocity
        max_v_inc = (max_w_inc*std::abs(v_inc))/std::abs(w_inc);
      }
      else
      {
        // overconstrain angular velocity
        max_w_inc = (max_v_inc*std::abs(w_inc))/std::abs(v_inc);
      }

      if (std::abs(v_inc) > max_v_inc)
      {
        // we must limit linear velocity
       cmd_vel->twist.linear.x  = last_cmd_vel.twist.linear.x  + sign(v_inc)*max_v_inc;
      }

      if (std::abs(w_inc) > max_w_inc)
      {
        // we must limit angular velocity
       cmd_vel->twist.angular.z = last_cmd_vel.twist.angular.z + sign(w_inc)*max_w_inc;
      }

      cmd_vel->header.stamp = ros::Time::now();
      smooth_vel_pub.publish(cmd_vel);
      last_cmd_vel = *cmd_vel;
    }
    else if (input_active == true)
    {
      // We already reached target velocity; just keep resending last command while input is active
      cmd_vel.reset(new geometry_msgs::TwistStamped());
      cmd_vel->twist = last_cmd_vel.twist;
      cmd_vel->header.stamp = ros::Time::now();
      smooth_vel_pub.publish(cmd_vel);

    }

    spin_rate.sleep();
  }
}

/**
 * Initialise from a nodelet's private nodehandle.
 * @param nh : private nodehandle
 * @return bool : success or failure
 */
bool VelocitySmoother::init(ros::NodeHandle& nh)
{
  // Dynamic Reconfigure
  ROS_WARN("Initializing the VelocitySmoother class ");
  dynamic_reconfigure_callback = boost::bind(&VelocitySmoother::reconfigCB, this, _1, _2);

  dynamic_reconfigure_server = new dynamic_reconfigure::Server<stamped_velocity_smoother::paramsConfig>(nh);
  dynamic_reconfigure_server->setCallback(dynamic_reconfigure_callback);

  // Optional parameters
  int feedback;
  nh.param("frequency",      frequency,     20.0);
  nh.param("quiet",          quiet,         quiet);
  nh.param("decel_factor",   decel_factor,   1.0);
  nh.param("robot_feedback", feedback, (int)NONE);

  if ((int(feedback) < NONE) || (int(feedback) > COMMANDS))
  {
    ROS_WARN("Invalid robot feedback type (%d). Valid options are 0 (NONE, default), 1 (ODOMETRY) and 2 (COMMANDS)",
             feedback);
    feedback = NONE;
  }

  robot_feedback = static_cast<RobotFeedbackType>(feedback);

  // Mandatory parameters
  if ((nh.getParam("speed_lim_v", speed_lim_v) == false) ||
      (nh.getParam("speed_lim_w", speed_lim_w) == false))
  {
    ROS_ERROR("Missing velocity limit parameter(s)");
    return false;
  }

  if ((nh.getParam("accel_lim_v", accel_lim_v) == false) ||
      (nh.getParam("accel_lim_w", accel_lim_w) == false))
  {
    ROS_ERROR("Missing acceleration limit parameter(s)");
    return false;
  }

  // Deceleration can be more aggressive, if necessary
  decel_lim_v = decel_factor*accel_lim_v;
  decel_lim_w = decel_factor*accel_lim_w;

  // Publishers and subscribers
  odometry_sub    = nh.subscribe("odometry",      1, &VelocitySmoother::odometryCB, this);
  current_vel_sub = nh.subscribe("robot_cmd_vel", 1, &VelocitySmoother::robotVelCB, this);
  raw_in_vel_sub  = nh.subscribe("raw_cmd_vel",   1, &VelocitySmoother::twistCB, this);
  smooth_vel_pub  = nh.advertise <geometry_msgs::TwistStamped> ("smooth_cmd_vel", 1);
  ROS_WARN(" initialized velocity smoother");
  return true;
}


/*********************
** Nodelet
**********************/

class VelocitySmootherNodelet : public nodelet::Nodelet
{
public:
  VelocitySmootherNodelet()  { }
  ~VelocitySmootherNodelet()
  {
    NODELET_DEBUG("Velocity Smoother : waiting for worker thread to finish...");
    vel_smoother_->shutdown();
    worker_thread_.join();
  }

  std::string unresolvedName(const std::string &name) const {
    size_t pos = name.find_last_of('/');
    return name.substr(pos + 1);
  }


  virtual void onInit()
  {
    ROS_WARN(" VelocitySmootherNodelet initialization");
    ros::NodeHandle ph = getPrivateNodeHandle();
    std::string resolved_name = ph.getUnresolvedNamespace(); // this always returns like /robosem/goo_arm - why not unresolved?
    std::string name = unresolvedName(resolved_name); // unresolve it ourselves
    NODELET_WARN_STREAM("Velocity Smoother : initialising nodelet...[" << name << "]");
    vel_smoother_.reset(new VelocitySmoother(name));
    if (vel_smoother_->init(ph))
    {
      NODELET_WARN_STREAM("Velocity Smoother : nodelet initialised [" << name << "]");
      worker_thread_ = boost::thread(boost::bind(&VelocitySmoother::spin, vel_smoother_));
    }
    else
    {
      NODELET_ERROR_STREAM("Velocity Smoother : nodelet initialisation failed [" << name << "]");
    }
  }

private:
  boost::shared_ptr<VelocitySmoother> vel_smoother_;
  boost::thread                       worker_thread_;
};

} // namespace stamped_velocity_smoother

PLUGINLIB_EXPORT_CLASS(stamped_velocity_smoother::VelocitySmootherNodelet, nodelet::Nodelet);
